device/cuda/include/cuda/seeding/seed_finding.hpp

```diff
@@ -96,56 +99,54 @@ struct seed_finding {
-            mid_bot_container, mid_top_container, m_mr);
+            m_seedfinder_config, g2, doublet_counter_container,
+            mid_bot_container, mid_top_container, m_mr.get());
 
         // triplet counting
-        traccc::cuda::triplet_counting(m_seedfinder_config, isp_container,
+        traccc::cuda::triplet_counting(m_seedfinder_config, g2,
                                        doublet_counter_container,
                                        mid_bot_container, mid_top_container,
-                                       triplet_counter_container, m_mr);
+                                       triplet_counter_container, m_mr.get());
 
         // triplet finding
         traccc::cuda::triplet_finding(
-            m_seedfinder_config, m_seedfilter_config, isp_container,
+            m_seedfinder_config, m_seedfilter_config, g2,
             doublet_counter_container, mid_bot_container, mid_top_container,
-            triplet_counter_container, triplet_container, m_mr);
+            triplet_counter_container, triplet_container, m_mr.get());
 
         // weight updating
-        traccc::cuda::weight_updating(m_seedfilter_config, isp_container,
+        traccc::cuda::weight_updating(m_seedfilter_config, g2,
                                       triplet_counter_container,
-                                      triplet_container, m_mr);
+                                      triplet_container, m_mr.get());
 
         // seed selecting
         traccc::cuda::seed_selecting(
-            m_seedfilter_config, isp_container, doublet_counter_container,
-            triplet_counter_container, triplet_container, seed_container, m_mr);
-
-        return seed_container;
+            m_seedfilter_config, g2, doublet_counter_container,
+            triplet_counter_container, triplet_container, seed_container,
+            m_mr.get());
     }
 
     private:
-    bool first_alloc;
     const seedfinder_config m_seedfinder_config;
     const seedfilter_config m_seedfilter_config;
     multiplet_estimator m_estimator;
     seed_filtering m_seed_filtering;
-
-    host_doublet_counter_container doublet_counter_container;
-    host_doublet_container mid_bot_container;
-    host_doublet_container mid_top_container;
-    host_triplet_counter_container triplet_counter_container;
-    host_triplet_container triplet_container;
-    host_seed_container seed_container;
-    vecmem::memory_resource* m_mr;
+    std::reference_wrapper<vecmem::memory_resource> m_mr;
+
+    // mutable internal objects for multiplets
+    mutable std::shared_ptr<std::mutex> mutex{std::make_shared<std::mutex>()};
+    mutable host_doublet_counter_container doublet_counter_container;
+    mutable host_doublet_container mid_bot_container;
+    mutable host_doublet_container mid_top_container;
+    mutable host_triplet_counter_container triplet_counter_container;
+    mutable host_triplet_container triplet_container;
+    mutable host_seed_container seed_container;
```

Comment:
> These are mutable because the operator() is const. Does it have to be? mutable is generally a bit of a code smell, and if we have this use-case, maybe algorithms should be allowed to have non-const operator()?

Isn't it related to thread-safe design? so I think explicitly declaring `mutable` objects is better while keeping `const` function.

> Also, can you explain why you need a shared_ptr for the mutex? Do you synchronize multiple instances of this class?

No. It's just one of my bad habits to use `shared_ptr` everywhere. I changed it `unique_ptr`

Reply:
This is probably fine. 

I just meant that generally, one should try to avoid having a `const` method which modifies `mutable` members, even if you use a mutex to make it thread-safe. This is more of a high level argument that if there's a possibility to simply have it non-const, and thereby communicate that it's not thread-safe on it's own and the caller has to ensure that it's called in a safe way, that is generally prefferable.

Anyway, let's keep the `mutable` here, if we want the `operator()()` to be `const`.